### DISCUSSION TITLE
About homing attack

### DISCUSSION THREAD

**Discussion by CharlieMalphas** [Techokami/SonicWorldsNext#34] [Q&A] [answered]
Thank you for your answer about Smooth Rotation today. Let me ask you one more question. I am thinking of adding a homing attack to Sonic's actions. If there is a way to add it, I would appreciate it if you could let me know.

**Comment by DimensionWarped** [comment, marked as answer]
That is a complex question with way more than one answer for how to approach it, but I'll get you started by telling you how you would change someone's air double-jump action into a double jump.

In the air state script (Air.gd), around like 27 there is the beginning of a big set of conditionals for what to do when a player presses jump while airborne (provided they haven't exhausted their air ability already). Let's say we want Knuckles to be the one to double jump instead of glide. We'd simply rip off the velocity setting code from lightning shield using that to replace the current parent.movement value and then remove the set_state below it. Optionally you could also play a sound of course.

To go from there to a drop dash, you could rip off the fire shield instead, just adding a slight downward component.

To get your homing attack, you'll probably need a homing attack state to use (similar to how Tails and Knuckles currently have flight and glide states respectively) that handles the actual homing aspect of the homing attack and the special bounce motion. You'll also need some new collision shape around your player that detects enemies entering. You'll need some heuristic for deciding which enemy inside of the collision should be targeted at any given time. Somewhere in your air action you'll want to check that collision mesh for any enemy objects (see the collision layers page on the wiki) that overlap it, apply your heuristic to select the target, apply some initial velocity determined by some other heuristic, and then swap your state to the homing attack state which should be able to adjust player velocity with the enemy's changing position.

Obviously, I haven't even begun to cover the various special visual effects and such you'd want for that, but I hope that's enough to get you started.
